core/controllers/feedback.py

```diff
@@ -33,9 +33,15 @@ class ThreadListHandler(base.BaseHandler):
 
     @acl_decorators.can_play_exploration
     def get(self, exploration_id):
-        self.values.update({
-            'threads': [t.to_dict() for t in feedback_services.get_all_threads(
-                exploration_id, False)]})
+        if constants.USE_NEW_SUGGESTION_FRAMEWORK:
+            self.values.update({
+                'threads': [t.to_dict() for t in feedback_services.get_threads(
```

Comment:
I don't understand what's going on here, sorry. The old code got all non-suggestion threads. This one gets both suggestion and non-suggestion threads, right? Isn't that a change in behaviour?

Also ... tests? There should be tests that detect this.

Reply:
I also don't understand what this has to do with the title of the PR ("suggestions subject not displayed").

Reply:
Sorry for not being clear. So each suggestion thread has a feedback thread linked to it. In the old framework, the suggestion model itself had a description field which was used to populate the subject field in the display. However, in the new framework, we are storing this data in the thread corresponding to the suggestion as its subject (This was done before too, the description was stored in the old suggestion as well as the thread subject). So now in the frontend, we get the suggestion data as well as it's linked thread and then derive the subject from the thread data. 

The reason the error came up was because the feedback thread listing only got the threads without suggestions (whereas as per the above logic, the threads linked to the suggestion were also expected by the frontend). So that's why we needed this change in behaviour.

Reply:
I'm sorry, I'm still confused.

I get that there are some differences in whether the description is stored in the Suggestion model or the FeedbackThread model. But two questions follow on from your explanation:

(a) How did the frontend get the suggestion threads before? Why don't we use that handler instead?
(b) Wouldn't frontend changes be needed to accommodate this change?

Reply:
(Also, I think there should still be tests.)

Reply:
For (a), what do you mean by "similarly"? Again, I'm looking at the old code, and there's no "True" case. My question was also about the suggestion threads, not the suggestion data. (This follows on from: "The old code got all non-suggestion threads. This one gets both suggestion and non-suggestion threads, right? Isn't that a change in behaviour?")

For (b) it sounds to me like your new threads are inconsistent with your old threads, and there is some sort of back-fix going on here. This just raises more questions.

Overall, I'm starting to become a bit skeptical of this change -- it seems hard to get conceptual clarity on what's actually going on, and I don't understand the big picture. In the beginning, before the migration, we have (F1) standalone feedback threads, (S1) feedback threads with suggestions attached to them. After the migration, we'll also have (F2) standalone feedback threads that are created after the migration, (S2) feedback threads corresponding to suggestions that were created after the migration.

Before the migration, the code in question fetched only F1. After the migration, the code in question fetches F2 and S2. You're saying (I think) that you need to get S2 now, whereas you didn't need to get S1 before. But isn't there some frontend handler that was only expecting F1, and that now expects both F and S? I need to see that change in order to approve this PR, since, in general, refactors should not cause a frontend/backend API to change.

Also, before the migration, S1 was created without has_suggestion=True. After the migration, S2 is created with has_suggestion=True. I guess it's clearly true that S should have has_suggestion=True, so is has_suggestion a newly-introduced field? (Basically, I don't understand why you're talking about "changing has_suggestion to true" when it should always have been true.)

I hope this explains the rationale behind my questions and introduces some terminology (F1, S1, F2, S2) that might be useful for talking about this. You might need to introduce F3, S3, etc. if there are additional time durations I'm not considering (e.g. between the faulty migration and the fixed migration).

Reply:
> (Basically, I don't understand why you're talking about "changing has_suggestion to true" when it should always have been true.)

This was a mistake on my part. I had missed this out initially and made the change in #5155. The field is not new, and was always there.

The corresponding frontend handler is this block which expects F2 & S2.
 https://github.com/oppia/oppia/blob/a3e3b9593808efe1c441ba11336c6a3c32bdba89/core/templates/dev/head/pages/exploration_editor/feedback_tab/ThreadDataService.js#L73

Reply:
Sorry -- for the first question, can you point to the line on the specific PR which shows the old/new diff? I need to see the actual change (not just the final code) -- it sounds like the frontend API changed and the backend API did not, and I need to understand why this issue wasn't caught then.

Also, I think I understand what happened re the second question now. What can be done to prevent such mistakes in the future?

Reply:
https://github.com/oppia/oppia/pull/5093/files#diff-4cceb961b44b004d4ba52ccba7aaa2d7R73 This PR introduced the frontend changes (the same line is linked). 

The situation was like this. I had missed out to set has_suggestion=True in this PR. So all suggestions created were created (when manually tested) with feedback threads that mentioned has_suggestion=False. At that point of time, the thread handler returned all non-suggestion feedback threads, and the threads we created (wrongly) fell in this category. So everything worked fine on manual testing too. On correcting has_suggestion to True in a later PR, I didn't think of the description being affected by the same. 

I am not sure of exactly what we should do to prevent these mistakes. Maybe we should strengthen the protractor suite for suggestion framework? I'll also try to be more careful while making such changes.

Reply:
Ok will try to simplify that code and add e2e tests for the same in this PR.  Yes, there is a connection between suggestions and feedback threads as every suggestion always has a feedback thread linked to it. I think we can combine these calls and filter out into the 2 groups. Will do the same. 

Thanks a lot!